pkg/smokescreen/listener.go

```diff
@@ -0,0 +1,61 @@
+package smokescreen
+
+import (
+	"net"
+	"sync"
+
+	"github.com/stripe/go-einhorn/einhorn"
+)
+
+// einhornListener is a net.Listener that will send an ACK to the einhorn
+// master the first time its Accept method is called.
+//
+// Its implementation leaves a short time window between the sending the ACK
+// and when a thread is actively blocked by an Accept syscall. Because there is
+// a gap here it is possible for the process to freeze/crash/etc and never
+// reach the Accept call. This may lead to sadness because of dropped
+// connections (if the accept queue overflows) or timeouts because nothing
+// realized this process died.
+//
+// The alternatives to work around this are complex, introduce concurrency and
+// create several new classes of errors. So I've erred on the side of
+// simplicity and am hoping that by moving the ACK as close to the Accept as
+// possible we'll avoid too much pain.
+type einhornListener struct {
+	net.Listener
+
+	firstAcceptLock sync.Mutex
+	firstAcceptOnce sync.Once
+
+	accept func() (net.Conn, error)
+}
+
+func (el *einhornListener) firstAccept() (conn net.Conn, err error) {
+	el.firstAcceptLock.Lock()
```

Comment:
You probably don't need this lock here; the docs say:

> Because no call to Do returns until the one call to f returns, if f causes Do to be called, it will deadlock.

Which means that the `el.firstAcceptOnce` is an implicit mutex until it has run at least once?